### PR TITLE
Added support to redis cluster storage, without transaction on opened…

### PR DESCRIPTION
**Pull request by felipeagger**
Added support to redis cluster storage, without redis.transaction (because it is not supported by cluster mode) on opened_at.setter

new storage class CircuitRedisClusterStorage.

### CODE REVIEW

File: src/pybreaker.py [hunk 1]
```diff
@@ -606,6 +606,159 @@ def _namespace(self, key):
...
+        Returns a datetime object of the most recent value of when the circuit
+        was opened.
+        """
+        try:
+            timestamp = self._redis.get(self._namespace('opened_at'))
+            if timestamp:
+                return datetime(*time.gmtime(int(timestamp))[:6])
+        except self.RedisError:
+            self.logger.error('RedisError', exc_info=True)
+            return None
```
Comment (danielfm) [inline src/pybreaker.py:731]: As you said, this is the only bit that changes from the current implementation.

Can't we find a way to add support for this without having to duplicate the whole class?

**Review by danielfm — COMMENTED** [review]
Thanks for the contribution.

File: src/pybreaker.py (hunk 1 above)
Comment (felipeagger) [reply src/pybreaker.py]: yes, I agree, I adjusted it by creating an extra parameter.

File: src/pybreaker.py [hunk 2]
```diff
@@ -587,14 +588,26 @@ def opened_at(self, now):
...
+
+            if self._cluster_mode:
+
+                current_value = self._redis.get(key)
                 next_value = int(calendar.timegm(now.timetuple()))
-                pipe.multi()
+
                 if not current_value or next_value > int(current_value):
-                    pipe.set(key, next_value)
+                    self._redis.set(key, next_value)
```
Comment (danielfm) [inline src/pybreaker.py:598]: That's it! Now the only thing I miss is a test case to exercise this code path here, to make sure it works as expected.

**Review by MarcosGomesKabum — APPROVED** [review]

**Comment by felipeagger** [comment]
TestCase Done! CircuitBreakerRedisTestCase.test_cluster_mode

**Comment by danielfm** [comment]
Thanks. The test here is failing in my env (Python 3.10.2), could you take a look?

```
======================================================================
FAIL: test_cluster_mode (tests.CircuitBreakerRedisTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/danielfm/Projects/personal/active/pybreaker/src/tests.py", line 741, in test_cluster_mode
    self.assertEqual(storage.opened_at, now)
AssertionError: datetime.datetime(2022, 3, 2, 17, 37, 25) != datetime.datetime(2022, 3, 2, 17, 37, 25, 863723)

----------------------------------------------------------------------
Ran 71 tests in 7.099s

FAILED (failures=1)
Test failed: <unittest.runner.TextTestResult run=71 errors=0 failures=1>
error: Test failed: <unittest.runner.TextTestResult run=71 errors=0 failures=1>
```

**Comment by felipeagger** [comment]
> Thanks. The test here is failing in my env (Python 3.10.2), could you take a look?
> 
> ```
> ======================================================================
> FAIL: test_cluster_mode (tests.CircuitBreakerRedisTestCase)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/danielfm/Projects/personal/active/pybreaker/src/tests.py", line 741, in test_cluster_mode
>     self.assertEqual(storage.opened_at, now)
> AssertionError: datetime.datetime(2022, 3, 2, 17, 37, 25) != datetime.datetime(2022, 3, 2, 17, 37, 25, 863723)
> 
> ----------------------------------------------------------------------
> Ran 71 tests in 7.099s
> 
> FAILED (failures=1)
> Test failed: <unittest.runner.TextTestResult run=71 errors=0 failures=1>
> error: Test failed: <unittest.runner.TextTestResult run=71 errors=0 failures=1>
> ```

Adjusted!
```
=================================================== test session starts ===================================================================================================================================================
platform linux -- Python 3.10.2, pytest-7.0.1, pluggy-1.0.0
rootdir: /.../pybreaker
collected 71 items                                                                                                                                                                                                                                                                                                        

src/tests.py .......................................................................                                                                                                                                                                                                                                [100%]

============================================ 71 passed, 140 warnings in 7.02s =============================================================================================================================================
```

**Comment by danielfm** [comment]
Thanks.

**Pull request merged**